### PR TITLE
Update Scenario0_Tests.xaml.cs

### PR DESCRIPTION
**Pull request by RuilinZhuIntel**
1. Added new test case MagInterference;
2. Added new test case StaticAccuracy;
3. Added new sensor Relative Orientation ;
4. Modified the CreateTestButton function: can set Button Font size in its argument, to handle the case that the test case name very long.
5. Added new function CalculateMagInterference.
6. Added some useful function such as innerProduct and norm, Angle4 and so on.

### CODE REVIEW

**Comment by jianliu617** [comment]
@jialzhu Could you please review this new pull request and merge if it's acceptable? Thank you.

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 1]
```diff
@@ -329,8 +337,14 @@ private void DisplayPrecondition()
                     instruction.Text = "Put device on a level surface, isolated from outside vibration.\n" +
                                        "Keep it in stationary state.";
                     break;
+                case "StaticAccuracy":
+                    instruction.Text = "Put device on a level surface with (Y axis) pointing to the magnet north.";
+                    break;
+                case "MagInterference":
+                    instruction.Text = "Please moving approximately 1G magnet pass stationary device in about 0.25m / s\n" +
+                    "you have 30 seconds to do the test.";
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: Suggest to change the instruction to: 
"Please move a magnet (~1G) pass the stationary device at a speed of approximately 0.25 m/s ."
"You have 30 seconds to complete the test."

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 2]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
+            double inner = 0;
+            inner += quaternionA[0] * quaternionB[0];
+            inner += quaternionA[1] * quaternionB[1];
+            inner += quaternionA[2] * quaternionB[2];
+            inner += quaternionA[3] * quaternionB[3];
+            return inner;
+        }
+        private double norm(double[] quaternion)
+        {
+            double SumOfSquares = 0;
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: rename to sumOfSquares

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 3]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
             eulerAngle[1] = Math.Atan2(sinr, cosr);
-
             // yaw (z-axis rotation)
             double siny = 2.0 * (quaternion.W * quaternion.Z + quaternion.X * quaternion.Y);
             double cosy = 1.0 - 2.0 * (quaternion.Y * quaternion.Y + quaternion.Z * quaternion.Z);
             eulerAngle[2] = Math.Atan2(siny, cosy);
-
             return eulerAngle;
         }
+        private double innerProduct(double[] quaternionA, double[] quaternionB)
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: rename to InnerProduct

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 4]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
+        {
+            double inner = 0;
+            inner += quaternionA[0] * quaternionB[0];
+            inner += quaternionA[1] * quaternionB[1];
+            inner += quaternionA[2] * quaternionB[2];
+            inner += quaternionA[3] * quaternionB[3];
+            return inner;
+        }
+        private double norm(double[] quaternion)
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: rename to Norm

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 5]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
+            foreach (double[] array in dataList)
+            {
+                angleSum += Angle4(array, expectedValue);
+            }
+            result = angleSum / dataList.Count;
+            if (result > 180) result = 360 - Math.Abs(result);
+            if (Math.Abs(result) >= 30 && Math.Abs(result) <= 60) result = Math.Abs(45 - result);
+            return result;
+        }
+        private async void staticAccuracyHandler()
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: rename to StaticAccuracyHandler

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 6]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
+            for (int i = 0; i < 15; i++)
+            {
 
+                instruction.Text = "Orientation Sensor" + " " + testType + " Test in progress...";
+                currentOrientationSensor.ReadingChanged -= OrientationSensorReadingChanged;
+                dataList.Clear();
+                for (int count = 10; count >= 0; count--)
+                {
+                    if (i == 0 || i == 5 || i == 10 || i == 14) instruction.Text = "You have" + " " + count + " " + 
+                            "seconds to Rotate the device on a level surface with its y axis pointing to magnetic north...";
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: "seconds to rotate the device on a level surface with its y axis pointing to magnetic north..."

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 7]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
+                    else if (i < 5) instruction.Text = "You have" + " " + count + " " + 
+                            "seconds to Rotate 90 degrees counterclockwise around the z axis and stop to a static position...";
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: "seconds to rotate the device 90 degrees counterclockwise around the z axis and stop at a static position..."

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 8]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
+                    else if (i < 10) instruction.Text = "You have" + " " + count + " " + 
+                            "seconds to Rotate 90 degrees counterclockwise around the x axis and stop to a static position...";
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: "seconds to rotate the device 90 degrees counterclockwise around the x axis and stop at a static position..."

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 9]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
+                    else instruction.Text = "You have" + " " + count + " " + 
+                            "seconds to Rotate 90 degrees counterclockwise around the y axis and stop to a static position...";
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: "seconds to rotate the device 90 degrees counterclockwise around the y axis and stop at a static position..."

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 10]
```diff
@@ -1435,20 +1558,96 @@ private double[] QuaternionToEulerAngle(SensorQuaternion quaternion)
...
+                for (int count = 5; count >= 0; count--)
+                {
+                    instruction.Text = "Sampling data for" + " " + count + " " + "seconds...";
+                    await Task.Delay(1000);
+                }
+                OriAngles.Add(OriAngleCalculate(i));
+            }
+            countdown = new Countdown(testLength[testType], testType);
+            startTime = DateTime.Now;
+            instruction.Text = testType + " Is ending...";
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: "is ending..."

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 11]
```diff
@@ -30,17 +30,20 @@ public enum Directions { left, right, up, down, nothing }
...
                 { "Drift", 60*15 },
-                { "PacketLoss", 60*5}
+                { "PacketLoss", 60*5},
+                { "StaticAccuracy", 5},
+                { "MagInterference", 30}
             }; // In seconds
         private List<int> indices;
         private string testType;
         private List<double[]> dataList;
+        private List<double> OriAngles;
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: rename to oriAngles

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 12]
```diff
@@ -329,8 +337,14 @@ private void DisplayPrecondition()
                     instruction.Text = "Put device on a level surface, isolated from outside vibration.\n" +
                                        "Keep it in stationary state.";
                     break;
+                case "StaticAccuracy":
+                    instruction.Text = "Put device on a level surface with (Y axis) pointing to the magnet north.";
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: "Put device on a level surface with its Y axis pointing to the magnet north."

File: sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs [hunk 13]
```diff
@@ -341,21 +355,18 @@ private void AddPivotItem(int sensorType, int index, int totalIndex)
...
-
             Button orientationTestButton = CreateTestButton("Orientation Test");
             Button frequencyTestButton = CreateTestButton("Frequency Test");
             Button offsetTestButton = CreateTestButton("Offset Test");
             Button jitterTestButton = CreateTestButton("Jitter Test");
             Button driftTestButton = CreateTestButton("Drift Test");
             Button packetLossTestButton = CreateTestButton("Packet Loss Test");
-       
+            Button staticAccuracyButton = CreateTestButton("Static Accuracy Test");
+            Button MagInterferenceButton = CreateTestButton("MagInterference Test", 28);
```
Comment (jialzhu) [inline sensors/Tools/SensorExplorer/code/SensorExplorer/Scenario0_Tests.xaml.cs]: rename to "magInterferenceButton"

**Comment by RuilinZhuIntel** [comment]
Hello jialzhu, thanks for the comments, I've renamed all the variables and function names you mentioned and changed some of the instructions based on your advice in the latest commit. Please have a review. Thanks!

**Comment by jialzhu** [comment]
Hi Ruilin,
The changes look good. I just merged it into the master branch.
Thanks,
Lisa

**Pull request merged**